File: apiome-rest/src/app/browse_facets.py
"""
Protocol / format facet vocabulary for the public browse directory (MFI-6.1).

The browse surfaces (``/v1/browse/*`` and the ``apiome-browse`` app) let a visitor narrow the
directory by **protocol** — the canonical :class:`~app.canonical_model.ApiParadigm` an artifact was
imported as (REST, RPC, event-driven, graph, data-schema, agent) — and by **specific format** (the
``source_format`` key an adapter recorded at import: ``openapi-3.1``, ``protobuf``, ``graphql``, …).

Both facet axes read the columns MFI-7.1 added to ``apiome.versions`` (``protocol`` /
``source_format``, backed by the partial facet indexes that migration created). This module owns the
two things SQL should not: **normalizing a caller-supplied filter value** to the stored vocabulary,
and **labelling** a stored value for display. Everything here is pure and side-effect free, so the
route layer and the tests can share one definition of "what is a valid protocol".

Labels reuse the existing registries rather than inventing a third one: protocol labels come from a
small table keyed by ``ApiParadigm`` values, and format labels come from the import-source registry
(which already names every adapter and the format keys it emits — "gRPC / Protobuf",
``openapi-3.1``), falling back to :mod:`~app.format_capability_registry`. A versioned key is
labelled from its adapter's name plus the version suffix ("OpenAPI 3.1", "Swagger 2.0"), and
anything still unrecognised degrades to the raw key — a facet never renders an empty chip and never
raises.
"""

from __future__ import annotations

import re
from functools import lru_cache
from typing import Dict, List, Mapping, Optional, Sequence, Tuple

from .canonical_model import ApiParadigm

__all__ = [
    "BROWSE_PROTOCOL_VALUES",
    "MAX_FACET_VALUE_LENGTH",
    "format_label",
    "normalize_format_filter",
    "normalize_protocol_filter",
    "protocol_label",
    "sort_format_counts",
    "sort_protocol_counts",
]


#: Longest facet value accepted from a caller. ``versions.source_format`` is ``VARCHAR(128)`` and
#: ``versions.protocol`` is ``VARCHAR(64)``, so anything longer cannot match a stored row; capping
#: keeps an oversized query string out of the SQL parameters entirely.
MAX_FACET_VALUE_LENGTH = 128


#: The protocol facet values, in the order they are presented. This is exactly the canonical
#: paradigm vocabulary — a browse facet must never offer a protocol an import cannot produce.
BROWSE_PROTOCOL_VALUES: Tuple[str, ...] = tuple(p.value for p in ApiParadigm)


#: Display labels for the protocol facet. Keyed by the stored ``versions.protocol`` value.
_PROTOCOL_LABELS: Mapping[str, str] = {
    ApiParadigm.REST.value: "REST",
    ApiParadigm.RPC.value: "RPC",
    ApiParadigm.EVENT.value: "Event-driven",
    ApiParadigm.GRAPH.value: "Graph",
    ApiParadigm.DATA_SCHEMA.value: "Data schema",
    ApiParadigm.AGENT.value: "Agent",
}


#: Spellings a caller may reasonably send for a protocol, mapped to the stored value. Punctuation
#: is stripped before lookup, so ``data-schema``/``data schema``/``dataschema`` all arrive as
#: ``dataschema`` here. Mirrors ``resolveCatalogProtocol`` in the UI's catalog format registry.
_PROTOCOL_ALIASES: Mapping[str, str] = {
    "rest": ApiParadigm.REST.value,
    "http": ApiParadigm.REST.value,
    "rpc": ApiParadigm.RPC.value,
    "grpc": ApiParadigm.RPC.value,
    "event": ApiParadigm.EVENT.value,
    "events": ApiParadigm.EVENT.value,
    "eventdriven": ApiParadigm.EVENT.value,
    "messaging": ApiParadigm.EVENT.value,
    "graph": ApiParadigm.GRAPH.value,
    "graphql": ApiParadigm.GRAPH.value,
    "dataschema": ApiParadigm.DATA_SCHEMA.value,
    "dataschemas": ApiParadigm.DATA_SCHEMA.value,
    "schema": ApiParadigm.DATA_SCHEMA.value,
    "data": ApiParadigm.DATA_SCHEMA.value,
    "agent": ApiParadigm.AGENT.value,
    "agentic": ApiParadigm.AGENT.value,
    "mcp": ApiParadigm.AGENT.value,
}


#: Splits a versioned format key into its base key and version suffix (``openapi-3.1`` →
#: ``openapi`` + ``3.1``). The version may itself be multi-part (``json-schema-2020-12`` →
#: ``json-schema`` + ``2020-12``), and it must be separated from the base, so ``hl7v2``,
#: ``iso20022`` and ``asn1`` — whole keys that merely end in digits — are left alone.
_VERSIONED_FORMAT_RE = re.compile(r"^(?P<base>.+?)[-_.](?P<version>\d+(?:[.-]\d+)*)$")


def _strip_punctuation(value: str) -> str:
    """Reduce a token to its lower-case alphanumerics (``Data-Schema`` → ``dataschema``)."""
    return re.sub(r"[^a-z0-9]", "", value.strip().lower())


def normalize_protocol_filter(value: Optional[str]) -> Optional[str]:
    """Normalize a caller-supplied protocol filter to the stored ``versions.protocol`` vocabulary.

    Matching is punctuation- and case-insensitive and accepts the common spellings a client might
    send (``data-schema``, ``event-driven``, ``graphql``). A value that is not a known protocol is
    returned in its normalized lower-case form rather than rejected, so an unknown filter yields an
    empty, well-formed result set instead of a 4xx — the same "filters never fail, they only narrow"
    contract the existing ``search``/``domain`` browse filters have.

    Args:
        value: The raw ``protocol`` query parameter, or None.

    Returns:
        The stored protocol value (e.g. ``data_schema``), a normalized unknown token, or None when
        the input is absent or blank.
    """
    if value is None:
        return None
    raw = value.strip().lower()
    if not raw:
        return None
    alias = _PROTOCOL_ALIASES.get(_strip_punctuation(raw))
    if alias is not None:
        return alias
    return raw[:MAX_FACET_VALUE_LENGTH]


def normalize_format_filter(value: Optional[str]) -> Optional[str]:
    """Normalize a caller-supplied format filter to the stored ``versions.source_format`` form.

    Stored format keys are already lower-case adapter/normalizer keys (``openapi-3.1``,
    ``protobuf``), so normalization is a trim + lower-case + length cap. As with
    :func:`normalize_protocol_filter`, an unrecognised key narrows to nothing rather than erroring.

    Args:
        value: The raw ``format`` query parameter, or None.

    Returns:
        The normalized format key, or None when the input is absent or blank.
    """
    if value is None:
        return None
    raw = value.strip().lower()
    if not raw:
        return None
    return raw[:MAX_FACET_VALUE_LENGTH]


def protocol_label(value: Optional[str]) -> str:
    """Human label for a stored protocol value (``data_schema`` → ``Data schema``).

    Args:
        value: The stored ``versions.protocol`` value.

    Returns:
        The display label, falling back to the raw value for a protocol outside the canonical
        paradigm vocabulary (which can only appear in data written before this vocabulary existed).
    """
    key = (value or "").strip().lower()
    return _PROTOCOL_LABELS.get(key, (value or "").strip())


@lru_cache(maxsize=1)
def _adapter_format_labels() -> Mapping[str, str]:
    """Map every adapter key and emitted format key to that adapter's label, built once.

    The import-source registry already names each adapter for its source card ("OpenAPI / Swagger",
    "gRPC / Protobuf") and declares which format keys it emits (``openapi-3.1``, ``swagger-2.0``, …),
    which is exactly the lookup a format facet needs. Reusing it means a newly registered adapter
    labels its facet chips with no change here.

    Returns:
        A lower-cased key → label mapping; empty if the registry cannot be loaded, in which case
        labelling falls back to the capability registry and then to the raw key.
    """
    try:
        from .import_source import describe_import_sources

        descriptors = describe_import_sources()
    except Exception:  # pragma: no cover - defensive; browse must never 500 over a label
        return {}
    labels: Dict[str, str] = {}
    for descriptor in descriptors:
        for key in (descriptor.key, *descriptor.formats):
            token = (key or "").strip().lower()
            if token:
                labels.setdefault(token, descriptor.label)
    return labels


def _capability_label(format_key: str) -> Optional[str]:
    """Look up a format key's label in the capability registry, or None when it has none.

    The registry answers for *every* key — an unregistered one echoes the key back as its label —
    so "no label" here means "the registry knows nothing about this key". Any registry failure is
    swallowed: a browse listing must not 500 because a label could not be resolved.
    """
    try:
        from .format_capability_registry import capability_for

        label = capability_for(format_key).label
    except Exception:  # pragma: no cover - defensive; the registry is pure data
        return None
    # The registry echoes an unregistered key back verbatim as its label; a *registered* key always
    # differs in case or spelling ("asyncapi" → "AsyncAPI"), so an exact match means "not known".
    if not label or label.strip() == format_key:
        return None
    return label


def _name_for_base(label: str, base: str) -> str:
    """Pick the part of a multi-name adapter label that names ``base``.

    One adapter can cover several named formats ("OpenAPI / Swagger"), so a versioned key must be
    labelled with the right half: ``swagger-2.0`` is "Swagger 2.0", not "OpenAPI 2.0". When no part
    matches, the first is used — still better than repeating every name before the version.

    Args:
        label: The adapter label, possibly of the form ``A / B``.
        base: The format key with its version suffix removed.

    Returns:
        The single format name to prefix the version with.
    """
    parts = [p.strip() for p in label.split("/") if p.strip()]
    if not parts:
        return label
    target = _strip_punctuation(base)
    for part in parts:
        if _strip_punctuation(part) == target:
            return part
    return parts[0]


def format_label(format_key: Optional[str]) -> str:
    """Human label for a stored source-format key (``openapi-3.1`` → ``OpenAPI 3.1``).

    Resolution order:

    1. The import-source registry's adapter label for the key verbatim — this covers every emitted
       format key, so ``protobuf`` becomes "gRPC / Protobuf".
    2. For a versioned key, the adapter label for the whole key (or, failing that, for its base)
       narrowed to the name that matches the base, plus the version: "Swagger 2.0", "AsyncAPI 3".
    3. The capability registry's label, which also answers for reviewed formats no adapter emits.
    4. The raw key, so an unknown format still renders a usable chip.

    Args:
        format_key: The stored ``versions.source_format`` value.

    Returns:
        The display label; never empty for a non-empty key.
    """
    key = (format_key or "").strip().lower()
    if not key:
        return ""
    adapter_labels = _adapter_format_labels()
    match = _VERSIONED_FORMAT_RE.match(key)
    if match:
        base, version = match.group("base"), match.group("version")
        label = adapter_labels.get(key) or adapter_labels.get(base) or _capability_label(base)
        if label:
            return f"{_name_for_base(label, base)} {version}"
    direct = adapter_labels.get(key) or _capability_label(key)
    if direct:
        return direct
    return format_key.strip()


def sort_protocol_counts(counts: Dict[str, int]) -> List[Tuple[str, str, int]]:
    """Order protocol facet counts by the canonical paradigm order.

    Args:
        counts: Stored protocol value → number of matching directory entries.

    Returns:
        ``(value, label, count)`` triples: the canonical paradigms first, in
        :data:`BROWSE_PROTOCOL_VALUES` order, then any non-canonical value alphabetically.
    """
    known = [v for v in BROWSE_PROTOCOL_VALUES if v in counts]
    unknown = sorted(v for v in counts if v not in _PROTOCOL_LABELS)
    return [(v, protocol_label(v), counts[v]) for v in [*known, *unknown]]


def sort_format_counts(counts: Dict[str, int]) -> List[Tuple[str, str, int]]:
    """Order format facet counts by popularity, then key, for a stable, useful chip row.

    Args:
        counts: Stored ``source_format`` key → number of matching directory entries.

    Returns:
        ``(value, label, count)`` triples, highest count first and ties broken by key so the order
        never depends on database row order.
    """
    ordered: Sequence[str] = sorted(counts, key=lambda k: (-counts[k], k))
    return [(v, format_label(v), counts[v]) for v in ordered]

File: apiome-browse/lib/browseFacets.ts
/**
 * Paradigm / format browse facets (MFI-6.1, #3753; paradigm vocabulary FMT-1.6, #5417).
 *
 * The directory now spans many API description formats, so browsing has to be narrowable by
 * **paradigm** — the canonical interaction style an artifact was imported as (REST, RPC,
 * event-driven, graph, data schema, agent) — and by **specific format** (`openapi-3.1`,
 * `protobuf`, `graphql`, …). Both values are recorded per published revision by the import
 * pipeline (MFI-7.1) and rolled up per project / organization by the queries in `lib/db/helper.ts`.
 *
 * This module is the framework-free half of that feature: the vocabulary, the label lookup, the
 * facet-count roll-up and the filter predicate. Keeping it free of React and of `pg` means the
 * behaviour is unit-testable (`lib/__tests__/browseFacets.test.ts`) and the chip components stay
 * purely presentational. It intentionally mirrors `app/browse_facets.py` in apiome-rest so the
 * public API and this app describe the same facet with the same words.
 *
 * ### Where the paradigm vocabulary comes from
 *
 * {@link BROWSE_PARADIGMS} is not typed here — it is the generated `FORMAT_PARADIGMS` list, which
 * is `ApiParadigm` in apiome-rest projected through `app.format_counts`. A hand-kept copy could
 * offer a paradigm no import produces, or miss one that was added; reading the generated list means
 * the facet gains a paradigm the moment the registry does, and the drift gate in
 * `apiome-rest/tests/test_format_counts.py` fails if the generated module is stale.
 *
 * ### Why entities still carry `protocols`
 *
 * `versions.protocol` is the stored column, and `lib/db/helper.ts` aggregates it under that name.
 * The column predates the canonical vocabulary the rest of the system settled on; renaming it is a
 * migration, not a facet change. So the *stored* name survives on {@link FacetedEntity} and the
 * mapping to the paradigm axis happens once, in {@link entityValues}, rather than being spread
 * through the components.
 */

import { FORMAT_PARADIGMS } from './generated/formatCounts';

/** One selectable value on a facet axis, with how many entries carry it. */
export interface BrowseFacetOption {
  /** The stored value, used as the filter token (e.g. `data_schema`, `openapi-3.1`). */
  value: string;
  /** Display label (e.g. `Data schema`, `OpenAPI 3.1`). */
  label: string;
  /** How many entries in the current scope carry this value. */
  count: number;
}

/** The currently selected facet values; `null` on an axis means "any". */
export interface BrowseFacetSelection {
  /** The selected canonical paradigm (e.g. `data_schema`), or null for "any". */
  paradigm: string | null;
  /** The selected source format key (e.g. `openapi-3.1`), or null for "any". */
  format: string | null;
}

/** Anything the facets can be computed over — a project or organization row. */
export interface FacetedEntity {
  /**
   * Distinct paradigms across the entry's listed versions, under the stored column's name
   * (`versions.protocol`). Read through {@link entityValues}, never directly by a component.
   */
  protocols?: string[] | null;
  /** Distinct source formats across the entry's listed versions. */
  formats?: string[] | null;
}

/** The two facet axes. */
export type BrowseFacetAxis = 'paradigm' | 'format';

/** The empty selection — nothing narrowed. */
export const NO_FACET_SELECTION: BrowseFacetSelection = { paradigm: null, format: null };

/**
 * The canonical paradigms, in canonical order — the generated projection of `ApiParadigm` in
 * apiome-rest. A facet built from this list can never offer a paradigm no import produces.
 */
export const BROWSE_PARADIGMS: readonly { id: string; label: string }[] = FORMAT_PARADIGMS;

const PARADIGM_LABELS: Readonly<Record<string, string>> = Object.fromEntries(
  BROWSE_PARADIGMS.map((p) => [p.id, p.label])
);

const PARADIGM_ORDER: Readonly<Record<string, number>> = Object.fromEntries(
  BROWSE_PARADIGMS.map((p, index) => [p.id, index])
);

/**
 * Display labels for the format keys the import adapters emit, keyed by the key with any trailing
 * version segment removed (`openapi-3.1` → `openapi`). A key the table does not know still renders
 * — as itself — so a newly added adapter never leaves a blank chip.
 */
const FORMAT_LABELS: Readonly<Record<string, string>> = {
  openapi: 'OpenAPI',
  swagger: 'Swagger',
  arazzo: 'Arazzo',
  asyncapi: 'AsyncAPI',
  cloudevents: 'CloudEvents',
  graphql: 'GraphQL',
  protobuf: 'gRPC / Protobuf',
  grpc: 'gRPC / Protobuf',
  connectrpc: 'Connect RPC',
  thrift: 'Thrift',
  capnproto: "Cap'n Proto",
  wit: 'WIT (WebAssembly)',
  flatbuffers: 'FlatBuffers',
  corbaidl: 'CORBA IDL',
  oncrpc: 'ONC RPC',
  xmlrpc: 'XML-RPC',
  openrpc: 'OpenRPC',
  smithy: 'Smithy',
  typespec: 'TypeSpec',
  raml: 'RAML',
  postman: 'Postman',
  'http-file': 'HTTP Request File',
  kong: 'Kong Declarative Config',
  'gateway-api': 'Gateway API HTTPRoute',
  odata: 'OData',
  wsdl: 'WSDL',
  wadl: 'WADL',
  discovery: 'Google Discovery',
  apiblueprint: 'API Blueprint',
  zosconnect: 'z/OS Connect',
  'json-schema': 'JSON Schema',
  jtd: 'JSON Type Definition',
  'k8s-crd': 'Kubernetes CRD',
  avro: 'Avro',
  xsd: 'XSD',
  dtd: 'DTD',
  relaxng: 'RELAX NG',
  asn1: 'ASN.1',
  cddl: 'CDDL',
  arrow: 'Apache Arrow',
  dbt: 'dbt Project',
  odcs: 'ODCS Data Contract',
  'kafka-connect': 'Kafka Connect Schema',
  'sql-ddl': 'SQL DDL',
  cobolcopybook: 'COBOL Copybook',
  fhir: 'FHIR',
  hl7v2: 'HL7 v2',
  edix12: 'EDI X12',
  iso20022: 'ISO 20022',
  iso8583: 'ISO 8583',
  fix: 'FIX',
  'llm-tools': 'LLM Tools',
  mcp: 'MCP Server Manifest',
};

/**
 * Splits a versioned format key into base + version (`openapi-3.1` → `openapi` / `3.1`). The
 * version may be multi-part (`json-schema-2020-12`) and must be separated from the base, so keys
 * that merely end in digits (`iso20022`, `hl7v2`, `asn1`) are left whole.
 */
const VERSIONED_FORMAT_RE = /^(.+?)[-_.](\d+(?:[.-]\d+)*)$/;

/**
 * Normalize a raw facet value to the stored form: trimmed and lower-cased.
 *
 * @param value A value from a row, a URL query, or a chip click.
 * @returns The normalized value, or `null` when it is absent or blank.
 */
export function normalizeFacetValue(value: string | null | undefined): string | null {
  const token = (value ?? '').trim().toLowerCase();
  return token ? token : null;
}

/**
 * Display label for a paradigm value.
 *
 * @param value The stored paradigm (e.g. `data_schema`).
 * @returns The label, falling back to the raw value for anything outside the canonical vocabulary.
 */
export function paradigmLabel(value: string): string {
  const key = normalizeFacetValue(value);
  if (!key) return '';
  return PARADIGM_LABELS[key] ?? value.trim();
}

/**
 * Display label for a source-format key (`openapi-3.1` → `OpenAPI 3.1`).
 *
 * A versioned key is labelled from its family name plus the version, so `openapi-3.0` and
 * `openapi-3.1` stay distinguishable as chips instead of collapsing to one word.
 *
 * @param formatKey The stored format key.
 * @returns The label; never empty for a non-empty key.
 */
export function formatLabel(formatKey: string): string {
  const key = normalizeFacetValue(formatKey);
  if (!key) return '';
  const direct = FORMAT_LABELS[key];
  if (direct) return direct;
  const match = VERSIONED_FORMAT_RE.exec(key);
  if (match) {
    const family = FORMAT_LABELS[match[1]];
    if (family) return `${family} ${match[2]}`;
  }
  return formatKey.trim();
}

/**
 * Display label for a value on either axis.
 *
 * @param axis Which facet the value belongs to.
 * @param value The stored value.
 * @returns The label for that axis.
 */
export function facetLabel(axis: BrowseFacetAxis, value: string): string {
  return axis === 'paradigm' ? paradigmLabel(value) : formatLabel(value);
}

/**
 * The values an entry carries on one axis, normalized and de-duplicated.
 *
 * This is the single place the paradigm axis is mapped onto the stored `protocols` field, so the
 * column's older name never reaches a component or a caller.
 */
function entityValues(entity: FacetedEntity, axis: BrowseFacetAxis): string[] {
  const raw = axis === 'paradigm' ? entity.protocols : entity.formats;
  if (!Array.isArray(raw)) return [];
  const seen = new Set<string>();
  for (const value of raw) {
    const token = normalizeFacetValue(typeof value === 'string' ? value : String(value ?? ''));
    if (token) seen.add(token);
  }
  return [...seen];
}

/**
 * Roll entries up into the chip row for one axis.
 *
 * An entry counts once per distinct value it carries, so a project published in both OpenAPI and
 * gRPC appears under both chips. Counts deliberately ignore the current facet selection — the row
 * answers "what else could I pick", which is what makes a facet navigable — matching the
 * `facets` block the `/v1/browse/*` API returns.
 *
 * @param entities The entries in scope (already narrowed by any text search).
 * @param axis Which axis to count.
 * @returns Options ordered as the API orders them: paradigms in canonical order, formats by
 *   descending count with ties broken by value.
 */
export function computeFacetOptions(
  entities: readonly FacetedEntity[],
  axis: BrowseFacetAxis
): BrowseFacetOption[] {
  const counts = new Map<string, number>();
  for (const entity of entities) {
    for (const value of entityValues(entity, axis)) {
      counts.set(value, (counts.get(value) ?? 0) + 1);
    }
  }

  const options = [...counts.entries()].map(([value, count]) => ({
    value,
    label: facetLabel(axis, value),
    count,
  }));

  if (axis === 'paradigm') {
    const rank = (value: string) => PARADIGM_ORDER[value] ?? BROWSE_PARADIGMS.length;
    return options.sort(
      (a, b) => rank(a.value) - rank(b.value) || a.value.localeCompare(b.value)
    );
  }
  return options.sort((a, b) => b.count - a.count || a.value.localeCompare(b.value));
}

/**
 * Keep only the entries matching the selection.
 *
 * The two axes compose with AND, and each is satisfied when the entry carries the value on *any*
 * of its versions — the same "at least one version" rule the API's SQL filter uses.
 *
 * @param entities The entries in scope.
 * @param selection The selected paradigm/format, either of which may be null for "any".
 * @returns The matching entries, in their original order.
 */
export function filterByFacets<T extends FacetedEntity>(
  entities: readonly T[],
  selection: BrowseFacetSelection
): T[] {
  const paradigm = normalizeFacetValue(selection.paradigm);
  const format = normalizeFacetValue(selection.format);
  if (!paradigm && !format) return [...entities];
  return entities.filter((entity) => {
    if (paradigm && !entityValues(entity, 'paradigm').includes(paradigm)) return false;
    if (format && !entityValues(entity, 'format').includes(format)) return false;
    return true;
  });
}

/**
 * Toggle a chip: selecting the active value clears the axis, anything else replaces it.
 *
 * @param selection The current selection.
 * @param axis The axis whose chip was clicked.
 * @param value The clicked value, or null to clear the axis.
 * @returns The next selection (a new object; the input is not mutated).
 */
export function toggleFacet(
  selection: BrowseFacetSelection,
  axis: BrowseFacetAxis,
  value: string | null
): BrowseFacetSelection {
  const next = normalizeFacetValue(value);
  const current = normalizeFacetValue(axis === 'paradigm' ? selection.paradigm : selection.format);
  const resolved = next && next === current ? null : next;
  return axis === 'paradigm'
    ? { ...selection, paradigm: resolved }
    : { ...selection, format: resolved };
}

/**
 * Whether anything is selected — the "Clear filters" affordance hangs off this.
 *
 * @param selection The current selection.
 * @returns True when at least one axis is narrowed.
 */
export function hasFacetSelection(selection: BrowseFacetSelection): boolean {
  return Boolean(normalizeFacetValue(selection.paradigm) || normalizeFacetValue(selection.format));
}

/**
 * A short human summary of the active selection, for empty-state copy.
 *
 * @param selection The current selection.
 * @returns Something like `REST · OpenAPI 3.1`, or an empty string when nothing is selected.
 */
export function describeFacetSelection(selection: BrowseFacetSelection): string {
  const parts: string[] = [];
  const paradigm = normalizeFacetValue(selection.paradigm);
  const format = normalizeFacetValue(selection.format);
  if (paradigm) parts.push(paradigmLabel(paradigm));
  if (format) parts.push(formatLabel(format));
  return parts.join(' · ');
}
